[Feature] Duplicate pool

esizer opened the pull request:
🤖 Resolves #6546 

## 👋 Introduction

Adds the ability for pool operators to duplicate their pools

## 🕵️ Details

Adds a new mutation `duplicatePoolAdvertisement`. This needs to pass `team.id` along with the `pool.id` since the policy requires it in the args. I thought about making a new `duplicate` policy, does that make more sense?

```php
public function duplicate(User $user, $request)
{
  $existing = Pool::findOrFail($request["id"]);
  
  // Confirm the user can create pools for the team
  if ($user->isAbleTo("create-team-pool", $existing->team)) {
    return true;
  } else {
    return Response::deny("Cannot create a pool for that team.");
  }
}
```

## 🧪 Testing

Assist reviewers with steps they can take to test that the PR does what it says it does.

1. Build `npm run build:fresh`
2. Login as `pool@test.com`
3. Edit a pool
4. Scroll to the bottom
5. Confirm duplicate button exists
6. Confirm it opens and duplicates the pool
7. Login as `admin@test.com`
8. Edit a pool
9. Confirm the duplicate button does not appear (only pool operators can duplicate right now?)

## 📸 Screenshot

![Screenshot 2023-05-23 151216](https://github.com/GCTC-NTGC/gc-digital-talent/assets/4127998/7000c8e2-0819-464c-8dc4-e89d68735df0)

vd1992 commented:
Microscopic issue I noticed only by logging in as `pool@test.com`, Pool Operator runs a different query at /admin/pools, it also needs the `updatedDate` field otherwise you get updated as null for all pools. `GetMePools` in this case. 
If you want to roll this in you can, otherwise I can split off an issue. 

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/40485260/53d1a3bf-db5f-4839-b7d8-423841fa8dfa)

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/40485260/8e04185a-f7f7-4533-823b-fa6369e7a497)

vd1992 commented on apps/web/src/pages/Pools/EditPoolPage/components/StatusSection.tsx:
I think the bit from the acceptance criteria regarding handling for unsaved changes was missed.

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/40485260/eb8a5c23-dd74-4004-bde0-7f27f41bdf5c)

esizer replied:
So, I have no idea how this can easily be done. Seems like the same ask (that was too convoluted to properly implement) was made for the application profile page. The problem is, our current way of determiniing unsaved changes is flawed and has issues:

 - https://github.com/GCTC-NTGC/gc-digital-talent/issues/5359#issue-1536728561
 - https://github.com/GCTC-NTGC/gc-digital-talent/issues/4981#issue-1471634956
 - https://github.com/GCTC-NTGC/gc-digital-talent/issues/5486#issue-1557508841

Then we would need to do this for 8 forms at once to block an action when a user can easily make the changes after duplicating? 🤔 Also, seems strange this would be the only action we block with unsaved changed on this page instead of also blocking:

 - Publishing
 - Extending
 - Archiving
 - Closing
 - Deleting

esizer commented:
> Microscopic issue I noticed only by logging in as `pool@test.com`, Pool Operator runs a different query at /admin/pools, it also needs the `updatedDate` field otherwise you get updated as null for all pools. `GetMePools` in this case. I noticed this as I was looking to see what happened for sorting of duplicated pools.
> 
> If you want to roll this in you can, otherwise I can split off an issue. Just letting you know first in case you had any thoughts.

We are actually planning on removing this query entirely so I may just leave as is 😅 

https://github.com/gctc-ntgc/gc-digital-talent/pull/issues/6109

patcon commented:
cc @tristan-orourke for feedback on policies
cc @Jerryescandon for feedback on flow of unsaved fields during dup

Jerryescandon commented:
I guess we can go ahead without the unsaved changes toast, it seems like a bigger issue altogether.

esizer commented:
> I guess we can go ahead without the unsaved changes toast, it seems like a bigger issue altogether.

We could also look into applying this block to publishing as well 🤔

Jerryescandon commented:
> > I guess we can go ahead without the unsaved changes toast, it seems like a bigger issue altogether.
> 
> We could also look into applying this block to publishing as well 🤔

I had tasked Elisa before she left to [re-design this page](https://www.figma.com/proto/tH8X3HEgXsQVGTa2eheH8P/Pool-Updates---Admin-(elisa)?page-id=0%3A1&type=design&node-id=5034-52683&viewport=12510%2C3611%2C0.42&scaling=min-zoom&starting-point-node-id=5034%3A46129) using the profile section cards to resolve this problem a while ago.

tristan-orourke reviewed (changes requested):
The current method of replicating the pool only copies fields in the Pool table. This misses any data from relationships - skills and screening questions.

esizer commented:
> I like the plan to add a `duplicate` method to the PoolPolicy.

👍 Added in [add pool duplicate policy](https://github.com/GCTC-NTGC/gc-digital-talent/pull/6675/commits/87ca0ecd0473cb95e77d468cadfac012688eae27)

> I don't quite understand what the issue is that @vd1992 found, but unless its a difficult fix lets try to resolve it. It may go away with #6109, but #6109 isn't in this sprint.

Pretty easy fix, added in [add updated at to pools index query](https://github.com/GCTC-NTGC/gc-digital-talent/pull/6675/commits/bd1e14110a67f9a1150660e53324b282f590695b)

> The current method of replicating the pool only copies fields in the Pool table. This misses any data from relationships - skills and screening questions.

🤦‍♀️ Not sure where my head was at while doing this. I knew `$model->replicate()` did not copy relationships but forgot to add them. Added in [update replicate with relationships](https://github.com/GCTC-NTGC/gc-digital-talent/pull/6675/commits/1d9ee59b127edbfa3e4d0e1e2819c435d7113cc9)

tristan-orourke reviewed (approved):
Looks good!
